src/main/java/uk/gov/hmcts/reform/laubackend/idam/repository/IdamLogonAuditFindLogonRepository.java

```diff
@@ -0,0 +1,147 @@
+package uk.gov.hmcts.reform.laubackend.idam.repository;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.hmcts.reform.laubackend.idam.domain.IdamLogonAudit;
+
+import java.sql.Timestamp;
+import java.util.LinkedList;
+import java.util.List;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.Query;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+@SuppressWarnings("unchecked")
+@Repository
+public class IdamLogonAuditFindLogonRepository {
+    private static final String SELECT_CRITERIA = """
+        SELECT ila.id, ila.user_id, ila.service, ila.login_state, ila.log_timestamp,
+               decrypt_value(ila.email_address, :encryptionKey) as email_address,
+               decrypt_value(ila.ip_address, :encryptionKey) as ip_address""";
+
+    private static final String FROM = "FROM idam_logon_audit ila";
+    private static final String WHERE = "WHERE";
+    private static final String TIMESTAMP_START_CRITERIA = "ila.log_timestamp >= :startTime";
+    private static final String TIMESTAMP_END_CRITERIA = "ila.log_timestamp <= :endTime";
+    private static final String USER_ID_CRITERIA = "ila.user_id = :userId";
+    private static final String EMAIL_CRITERIA = """
+        encode(hmac(:emailAddress, :encryptionKey , 'sha256'), 'hex')
+        = ila.email_address_mac""";
+
+    @PersistenceContext
+    private final EntityManager entityManager;
+
+    @Autowired
+    public IdamLogonAuditFindLogonRepository(final EntityManager entityManager) {
+        this.entityManager = entityManager;
+    }
+
+    @Transactional(readOnly = true)
+    public Page<IdamLogonAudit> findIdamLogon(final String userId,
+                                              final String emailAddress,
+                                              final Timestamp startTime,
+                                              final Timestamp endTime,
+                                              final String encryptionKey,
+                                              final Pageable pageable) {
+        final List<String> queryParts = new LinkedList<>();
+
+        queryParts.add(SELECT_CRITERIA);
+        queryParts.add(FROM);
+
+        List<String> searchParts = createSearchCriteria(userId, emailAddress, startTime, endTime);
```

Comment:
You can create searchCriteria method for searchParts, so that can be reused in count query too.

Reply:
Good point, addressed.